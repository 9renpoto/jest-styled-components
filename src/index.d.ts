declare namespace jest {
  interface Matchers<R> {
    toHaveStyleRule(property: string, value: string): R;
  }
}
