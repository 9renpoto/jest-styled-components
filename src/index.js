const toHaveStyleRule = require('./toHaveStyleRule')
const styleSheetSerializer = require('./styleSheetSerializer')
const styleSheet = require('styled-components/lib/models/StyleSheet')
const { isOverV2, isServer } = require('./utils')

if (isOverV2()) {
  styleSheet.default.reset(isServer())
}

expect.addSnapshotSerializer(styleSheetSerializer)
expect.extend({ toHaveStyleRule })
