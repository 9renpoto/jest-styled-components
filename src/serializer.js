const css = require('css')
const styleSheet = require('styled-components/lib/models/StyleSheet')

const getClassNames = (node, classNames) => {
  if (node.children && node.children.reduce) {
    classNames = node.children.reduce((acc, child) => (
      acc.concat(getClassNames(child, acc))
    ), classNames)
  }
  if (node.props && node.props.className) {
    return classNames.concat(node.props.className.split(' '))
  }
  return []
}

const getStyles = (classNames) => {
  const styles = styleSheet.rules().map(rule => rule.cssText).join('\n')
  const ast = css.parse(styles)
  ast.stylesheet.rules = ast.stylesheet.rules.filter(rule => (
    rule.type === 'rule' && classNames.includes(rule.selectors[0].substring(1))
  ))

  return css.stringify(ast)
}

const serializer = {

  test(val) {
    return !val.withStyles && val.$$typeof === Symbol.for('react.test.json')
  },

  print(val, print) {
    const classNames = getClassNames(val, [])
    const styles = getStyles(classNames)
    val.withStyles = true

    return `${styles}\n\n${print(val)}`
  },

}

module.exports = serializer
