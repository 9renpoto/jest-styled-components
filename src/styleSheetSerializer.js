const css = require('css')
const { getCSS } = require('./utils')

const getClassNames = node => {
  const classNames = []

  if (node.children) {
    node.children
      .slice()
      .reverse()
      .forEach(child =>
        Array.prototype.unshift.apply(classNames, getClassNames(child))
      )
  }

  if (node.props && node.props.className) {
    Array.prototype.unshift.apply(classNames, node.props.className.split(/\s/))
  }

  return classNames
}

const includesClassNames = (classNames, selectors) =>
  classNames.some(className =>
    selectors.some(selector => selector.indexOf(className) > -1)
  )

const filterRules = classNames => rule =>
  rule.type === 'rule' &&
  includesClassNames(classNames, rule.selectors) &&
  rule.declarations.length

const getAtRules = (ast, filter) =>
  ast.stylesheet.rules
    .filter(rule => rule.type === 'media' || rule.type === 'supports')
    .reduce((acc, atRule) => {
      atRule.rules = atRule.rules.filter(filter)

      if (atRule.rules.length) {
        return acc.concat(atRule)
      }

      return acc
    }, [])

const getStyle = classNames => {
  const ast = getCSS()
  const filter = filterRules(classNames)
  const rules = ast.stylesheet.rules.filter(filter)
  const atRules = getAtRules(ast, filter)

  ast.stylesheet.rules = rules.concat(atRules)

  return css.stringify(ast)
}

const replaceClassNames = (classNames, style, code) => {
  let index = 0
  return classNames.reduce((acc, className) => {
    if (style.indexOf(className) > -1) {
      return acc.replace(new RegExp(className, 'g'), `c${index++}`)
    }

    return acc.replace(
      new RegExp(`(className="[^"]*)${className}\\s?([^"]*")`, 'g'),
      '$1$2'
    )
  }, `${style}${style ? '\n\n' : ''}${code}`)
}

const styleSheetSerializer = {
  test(val) {
    return (
      val && !val.withStyle && val.$$typeof === Symbol.for('react.test.json')
    )
  },

  print(val, print) {
    val.withStyle = true

    const classNames = getClassNames(val)
    const style = getStyle(classNames)
    const code = print(val)

    return classNames.length ? replaceClassNames(classNames, style, code) : code
  },
}

module.exports = styleSheetSerializer
