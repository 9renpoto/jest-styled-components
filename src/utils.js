const css = require('css')
const { ServerStyleSheet } = require('styled-components')
const StyleSheet = require('styled-components/lib/models/StyleSheet')

const STYLE_TAGS_REGEXP = /<style[^>]*>([^<]*)</g

const isOverV2 = () => Boolean(ServerStyleSheet)

const isServer = () => typeof document === 'undefined'

const resetStyleSheet = () => {
  if (isOverV2()) {
    StyleSheet.default.reset(isServer())
  }
}

const getStyle = html => {
  let style = ''
  let matches

  while ((matches = STYLE_TAGS_REGEXP.exec(html)) !== null) {
    style += matches[1].trim()
  }

  return style
}

const getCSS = () => {
  let style

  if (isOverV2()) {
    if (isServer()) {
      style = getStyle(new ServerStyleSheet().getStyleTags())
    } else {
      style = getStyle(StyleSheet.default.instance.toHTML())
    }
  } else {
    style = StyleSheet.rules().map(rule => rule.cssText).join('\n')
  }

  return css.parse(style)
}

module.exports = {
  resetStyleSheet,
  getStyle,
  getCSS,
}
