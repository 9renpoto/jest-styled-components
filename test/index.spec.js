import React from 'react'
import renderer from 'react-test-renderer'
import styled from 'styled-components'
import { matcher, serializer } from '../src'

expect.addSnapshotSerializer(serializer)
expect.extend(matcher)

const Title = styled.h1`
  font-size: 1.5em;
  text-align: center;
  color: palevioletred;
`

const Wrapper = styled.section`
  padding: 4em;
  background: papayawhip;
`

test('it works', () => {
  const tree = renderer.create(
    <Wrapper>
      <Title>Hello World, this is my first styled component!</Title>
    </Wrapper>,
  ).toJSON()

  expect(tree).toMatchStyledComponentsSnapshot()
})
