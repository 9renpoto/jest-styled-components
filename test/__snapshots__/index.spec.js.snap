// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`enzyme 1`] = `
.cfRhKF {
  padding: 4em;
  background: papayawhip;
}

<section
  className="cfRhKF"
>
  <styled.h1>
    Hello World, this is my first styled component!
  </styled.h1>
</section>
`;

exports[`react-test-renderer 1`] = `
.cfRhKF {
  padding: 4em;
  background: papayawhip;
}

.kZQsGZ {
  font-size: 1.5em;
  text-align: center;
  color: palevioletred;
}

<section
  className="cfRhKF"
>
  <h1
    className="kZQsGZ"
  >
    Hello World, this is my first styled component!
  </h1>
</section>
`;
